<a href="https://colab.research.google.com/github/slover1126/slover1126/blob/main/%ED%95%B8%EC%A6%88%EC%98%A8_%EB%A8%B8%EC%8B%A0%EB%9F%AC%EB%8B%9D_5%EC%9E%A5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**5장 서포트 벡터 머신**

**5.1 선형 SVM 분류**



*   서포트 벡터 머신: 분류의 경계선을 찾되, 그 경계 주변의 공간(마진)을 가장 크게 만드는 선을 찾는 모델
*   이때, 경계에 위치한 샘플들(서포트 벡터)들에 의해 경계가 결정이 됨(다른 샘플은 영향X)
* SVM은 특성의 스케일에 민감하기 때문에 StandardScaler를 사용하는것이 좋음 (한쪽의 단위가 너무 크면 왜곡이 발생)


**5.1.1 소프트 마진 분류**



*   **하드 마진 분류**:모든 샘플이 마진(경계 도로 폭) 밖으로 존재하게 분류함.  이상치가 있는경우 결정경계 계산이 안되거나 결정경계가 이상하게 생길 수 있다는 문제 발생
*   따라서 **소프트 마진 분류** 을 통해서 도로의 폭을 넓게 유지하는 것과 마진 오류(1번 클래스 데이터가 2번 클래스 쪽 결정경계 안에 위치하거나 마진 안에 있는 경우) 사이의 적절한 균형을 찾는것이 중요함
* 이때 **하이퍼파라미터 'C'** 가 사용됨. C를 줄이면 도로가 커지고 마진오류 발생이 증가함. 반대로 늘리면 포함되는 서포트 벡터가 감소해서 마진오류 발생이 감소함. 다만 각각 과소적합, 과대적합 발생 위험을 내포함(모델이 단순/복잡 해지므로)



In [2]:
from sklearn.datasets import load_iris
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import LinearSVC

iris = load_iris(as_frame=True)
X = iris.data[["petal length (cm)", 'petal width (cm)']].values
y = (iris.target == 2) #iris-virginica

#스케일링까지 한번에 파이프라인으로 구조화
svm_clf = make_pipeline(StandardScaler(), LinearSVC(C=1, dual = True, random_state=42))

svm_clf.fit(X,y)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('linearsvc', LinearSVC(C=1, dual=True, random_state=42))])

In [3]:
X_new = [[5.5,1.7], [5.0, 1.5]]
svm_clf.predict(X_new)
#첫 번째 꽃은 True(iris-virginica)로 분류되지만 2번째는 False(iris-virginica이 아님)으로 분류됨

array([ True, False])

In [4]:
svm_clf.decision_function(X_new)
# array의 첫번째 양수->True로 예

array([ 0.66163411, -0.22036063])

SVM에서 decision_function이 뱉는 숫자는 기준선이 아주 명확합니다.

0 (중앙선): 결정 경계 딱 그 자체입니다.

양수 ($> 0$): True (1번 클래스, iris-virginica)로 예측합니다.

음수 ($< 0$): False (0번 클래스, iris-virginica 아님)로 예측합니다.

+1 과 -1 (가드레일/마진): 서포트 벡터들이 위치한 도로 양쪽 가장자리입니다.

+1보다 크거나 -1보다 작으면: 도로 밖 안전지대에 잘 들어간 데이터입니다.

-1과 +1 사이의 값: 도로(마진) 내부에 아슬아슬하게 들어와 있는 데이터입니다.

즉, array([ 0.66163411, -0.22036063]) 에서 첫번째 원소는 X_new의 첫번째 원소에 대한 결정경계와의 거리, 두번째 원소는 X_new의 두번째 원소에 대한 결정경계의 거리이고, 둘 다 -1~1 사이의 값을 가지므로 마진 안의 데이터라고 할 수 있다.
다만, 첫번째는 양수이므로 True, 두번째는 음수이므로 False로 분류된다.